### ISSUE TITLE
[BUG]Seren Error when Refreshing Item Metadata & Refresh not working

### ISSUE THREAD

**Issue by Dealz-416LS** [nixgates/plugin.video.seren#585]
**Logs**
[kodi.log](https://github.com/nixgates/plugin.video.seren/files/6218220/kodi.log)

**Describe the bug**
Currently not able to properly refresh an item to update the Episode Thumb Artwork etc for individual episodes, Only way is to do a complete clear/refresh of metadata in Trakt Sync Tools.

When using `Context Menu > Trakt Manager > Refresh Item Metadata` on a Season item the item disappears along with a Seren error, navigating back to the item only the Season Artwork gets updated (Episode thumb artwork etc doesn't update.)

When using `Context Menu > Trakt Manager > Refresh Item Metadata` on the Show or Episode item it doesn't give an error but it doesn't refresh anything.

NOTE: Navigating back to the show item after a 'Refresh Item Metadata' on it gives a Seren Error.

**To Reproduce**
 `Context Menu > Trakt Manager > Refresh Item Metadata` on a season.

**Expected behavior**
Season item shouldn't disappear and the season should refresh it's metadata (Poster artwork, Episode artwork etc). 

**Screenshots**
N/A

**Kodi Version (please complete the following information):**
 - Kodi Version 19.0
 - Seren Version 2.0.18+Matrix. 1

**Additional context**
N/A

**Comment by Nuklear92** [comment]
Totally experiencing this same issue, not only that. But normally as Seren does it's "background syncs with Trakt I believe it should also automatically update titles that have old cached metadata info like, missing artwork, plot, cast, etc. the background Trakt syncs happening with Seren just removes or adds titles, Episodes, etc. only with the metadata info available at that point but no further updates happening. 

Example; I have personally clicked on new movie and tv shows titles released or about to release and at the time of checking them out they have tons of metadata info missing as well as artwork or backdrops. After a whole week and many "background" Trakt syncs with Seren I go back and check on those same titles and only they're still showing the same outdated metadata info and missing artworks and I go double checking on Trakt website which they do overnight metadata refreshes on their titles and these titles might have been updated with artwork, cast members as well as even an available trailer that previously was missing. However in Seren even after many syncs those titles remain static with old cached info, UNLESS you manually go and rebuild database. If you ask me, I believe we shouldn't be doing that too often specially with the amount of server resources that takes from Trakt. This metadata updates in titles should happen automatically with those minor background Trakt syncs Seren does every so often. 

I really hope this issue gets attention from the Devs working in this app cause it's a bit annoying Seren is showing outdated information on anything you click on and on a later day or week you come back and check and everything still the same as the first time you checked regardless of info being updated in the official Trakt website.

**Comment by drinfernoo** [comment]
Is this fixed as of 2.1.0?

**Comment by Nuklear92** [comment]
> Is this fixed as of 2.1.0?

Yes, you can close this one. 🙂

**Comment by Dealz-416LS** [comment]
> Is this fixed as of 2.1.0?

I haven't fully tested the Refresh Item metadata (context menu). 

So far it's worked when I used it on a episode item to refresh episode artwork.

However when I used it on a season item I did get a Seren error like my original report.

**Comment by Nuklear92** [comment]
> > Is this fixed as of 2.1.0?
> 
> I haven't fully tested the Refresh Item metadata (context menu). 
> 
> So far it's worked when I used it on a episode item to refresh episode artwork.
> 
> However when I used it on a season item I did get a Seren error like my original report.
> 
> I haven't tested on a show item.
> 
> 

I haven't tested the manual Refresh Item Metadata option in Context Menu - Trakt Manager myself much after the latest update, but for the other issue of Seren updating metadata in the background for titles missing artwork, cast, plot, etc. it's been working good so far and I had seen it updating some of my UpNext episodes and other movie titles that were missing either Trailers, Backdrops, and plots.

**Comment by drinfernoo** [comment]
If you guys could test that this is working for all items (movie, show, season, and episode), that would be fantastic. "Working" here would mean no error, and *if* there is updated metadata, that it is updated accordingly. If no new metadata is present, and nothing changes, that would still be acceptable.

**Comment by Nuklear92** [comment]
> If you guys could test that this is working for all items (movie, show, season, and episode), that would be fantastic. "Working" here would mean no error, and *if* there is updated metadata, that it is updated accordingly. If no new metadata is present, and nothing changes, that would still be acceptable.

I'll try testing tomorrow, however one thing to keep in mind is that even if **Refresh Item Metadata** doesn't do anything on a movie, show, season or episode. That doesn't automatically means it doesn't work, it could be that there was a background metadata refresh by Seren in the background recently and it pulled all of the updated metadata into the movie, show, season or episode from Trakt or anywhere else where it gathers metadata info with the new update. 

So, basically by hitting manually the **Refresh Item Metadata** under the context menu - Trakt manager won't do anything if all the info available has been updated. This is why that feature will be a bit harder to see if it's working or not individually from what Seren is now doing automatically in the background every so often.

**Comment by drinfernoo** [comment]
@Nuklear92 Yes, I mentioned this.

**Comment by Nuklear92** [comment]
@drinfernoo So, I been testing the manual Refresh Item Metadata on some movies, shows, seasons and individual episodes. However, I can safely say that the automatic background metadata refresh updates introduced on the latest Seren build is working wonderfully. Yesterday I updated some movie titles that were missing backdrops, cast info, among other info in tmdb. And just recently when I entered the Trakt list that had those titles I edited I came across that all of them had been updated with exactly all the info I entered and artworks, backdrops. So, in that regard I can honestly say whatever the devs did for Seren to do and refresh metadata on titles regularly is working reliably as of now. 

So, going back to the manual Refresh Item Metadata situation, I can't see to make it work or to see it work since Seren is already updating metadata in the background. It's hard to come across a title with outdated info of any kind now that requires you to manually check and use that feature found under the Trakt Manager in the Context Menu. One thing for sure I'd found out is on some TV Shows when you go and do a manual Refresh Item Metadata on the Season, not episodes or the title itself. It triggers a Seren Error which I manage to take a screenshot of it. Example TV Show titles I seen this happen are, **The Irregulars** and **Tribes of Europa**. I didn't tried more, but I'm sure there could be many others, I say on some titles because it doesn't trigger any errors when I do it on Season from titles like **Superman and Lois**, **Loki**.

![image](https://user-images.githubusercontent.com/76958541/122236340-6fb7ec80-ce8c-11eb-80cf-4ab63e29e2d4.png)

**Comment by drinfernoo** [comment]
I can reproduce... but I get a different error:
```python
2021-06-23 07:23:08.245 T:23080   ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'TypeError'>
                                                   Error Contents: 'NoneType' object is not subscriptable
                                                   Traceback (most recent call last):
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\seren.py", line 20, in <module>
                                                       router.dispatch(g.REQUEST_PARAMS)
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\modules\router.py", line 307, in dispatch
                                                       tvshowMenus.Menus().show_seasons(action_args)
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\gui\tvshowMenus.py", line 489, in show_seasons
                                                       self.list_builder.season_list_builder(args["trakt_id"], no_paging=True)
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\modules\list_builder.py", line 38, in season_list_builder
                                                       shows.TraktSyncDatabase().get_season_list(show_id, **params),
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\modules\guard_decorators.py", line 19, in wrapper
                                                       return func(*args, **kwargs)
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\database\trakt_sync\shows.py", line 267, in get_season_list
                                                       self._try_update_seasons(trakt_show_id, trakt_id)
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\modules\guard_decorators.py", line 19, in wrapper
                                                       return func(*args, **kwargs)
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\database\trakt_sync\shows.py", line 692, in _try_update_seasons
                                                       self._format_seasons(seasons_to_update)
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\modules\guard_decorators.py", line 38, in wrapper
                                                       return func(*args, **kwargs)
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\database\trakt_sync\shows.py", line 579, in _format_seasons
                                                       formatted_items = self._format_objects(self._identify_seasons_to_update(list_to_update))
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\database\trakt_sync\shows.py", line 484, in _format_objects
                                                       return self.metadataHandler.format_db_object(updated_items)
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\modules\metadataHandler.py", line 158, in format_db_object
                                                       return [self.format_meta(i) for i in db_object]
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\modules\metadataHandler.py", line 158, in <listcomp>
                                                       return [self.format_meta(i) for i in db_object]
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\modules\metadataHandler.py", line 176, in format_meta
                                                       self._apply_best_fit_meta_data(
                                                     File "C:\Users\drinf\Desktop\kodi\Matrix 19.0 Testing\portable_data\addons\plugin.video.seren\resources\lib\modules\metadataHandler.py", line 277, in _apply_best_fit_meta_data
                                                       media_type = trakt_data["info"]["mediatype"]
                                                   TypeError: 'NoneType' object is not subscriptable
                                                   -->End of Python script error report<--
```
We'll look at in dev.

**Comment by Dealz-416LS** [comment]
@drinfernoo This is what i get, Looks the same as your error.

```
2021-06-23 15:45:04.846 T:32533   ERROR <general>: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                                    - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                                   Error Type: <class 'TypeError'>
                                                   Error Contents: 'NoneType' object is not subscriptable
                                                   Traceback (most recent call last):
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/seren.py", line 20, in <module>
                                                       router.dispatch(g.REQUEST_PARAMS)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/modules/router.py", line 307, in dispatch
                                                       tvshowMenus.Menus().show_seasons(action_args)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/gui/tvshowMenus.py", line 485, in show_seasons
                                                       self.list_builder.season_list_builder(args["trakt_id"], no_paging=True)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/modules/list_builder.py", line 38, in season_list_builder
                                                       shows.TraktSyncDatabase().get_season_list(show_id, **params),
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/modules/guard_decorators.py", line 19, in wrapper
                                                       return func(*args, **kwargs)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/database/trakt_sync/shows.py", line 267, in get_season_list
                                                       self._try_update_seasons(trakt_show_id, trakt_id)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/modules/guard_decorators.py", line 19, in wrapper
                                                       return func(*args, **kwargs)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/database/trakt_sync/shows.py", line 680, in _try_update_seasons
                                                       self._format_seasons(seasons_to_update)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/modules/guard_decorators.py", line 38, in wrapper
                                                       return func(*args, **kwargs)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/database/trakt_sync/shows.py", line 567, in _format_seasons
                                                       formatted_items = self._format_objects(self._identify_seasons_to_update(list_to_update))
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/database/trakt_sync/shows.py", line 472, in _format_objects
                                                       return self.metadataHandler.format_db_object(updated_items)
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/modules/metadataHandler.py", line 158, in format_db_object
                                                       return [self.format_meta(i) for i in db_object]
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/modules/metadataHandler.py", line 158, in <listcomp>
                                                       return [self.format_meta(i) for i in db_object]
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/modules/metadataHandler.py", line 176, in format_meta
                                                       self._apply_best_fit_meta_data(
                                                     File "/storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/plugin.video.seren/resources/lib/modules/metadataHandler.py", line 277, in _apply_best_fit_meta_data
                                                       media_type = trakt_data["info"]["mediatype"]
                                                   TypeError: 'NoneType' object is not subscriptable
                                                   -->End of Python script error report<--
```

Also added short video clip of what happens when the error happens (item no longer displayed).


https://user-images.githubusercontent.com/54026944/123120056-ef6d2a80-d43b-11eb-8217-9d93f024dd1d.mp4

**Comment by drinfernoo** [comment]
We've identified the issue and are working on it :+1:

**Comment by drinfernoo** [comment]
Actually, we've fixed it :+1: Next update.

**Comment by drinfernoo** [comment]
https://github.com/nixgates/plugin.video.seren/issues/628

**Comment by Dealz-416LS** [comment]
No longer getting the Seren error I reported, I tested a few items.

However I wasn't able to test the actual refresh metadata functionality, I think for now this issue can be closed since I have no experience with 2.1.2 that the metadata refresh isn't working, so for now I assume it does.

**Comment by Nuklear92** [comment]
I can also confirm after trying it on two tv show titles that previously triggered an error when doing the manual refresh on the Season, now are not triggering errors which leads to believe the feature could be working.
